feat: refactor hgvsTools and support `c.` translation

larrybabb opened the pull request:
Address #365 

Begin refactoring the Translator logic to aggregate the functions and logic related to `hgvs` and provide support for `c.` from and to translation.  This should be step in the direction of re-thinking the organization of how the Translators are architected.

jsstevenson commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,143 @@
-    def __init__(self, uta_conn):
+    Attributes:
+        parser (hgvs.parser.Parser): The HGVS parser object.
+        uta_conn: The UTA (Unified Transcriptome and Annotation) connection object.
```
nitpick: it's "Universal Transcript Archive"

larrybabb replied:
fixed

jsstevenson commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -64,39 +56,49 @@ def __init__(
-        # prefix accession with namespace
-        sequence = coerce_namespace(sv.ac)
-        return self._get_refget_accession(sequence)
+        prefix_dict = {
+            "refseq:NM_": "n",
+            "refseq:NC_012920": "m",
+            "refseq:NG_": "g",
+            "refseq:NC_00": "g",
+            "refseq:NW_": "g",
+            "refseq:NR_": "n",
+            "refseq:NP_": "p",
+            "refseq:XM_": "n",
+            "refseq:XR_": "n",
+            "refseq:XP_": "p",
+            "GRCh": "g",
+        }
```
double checking: this mapping isn't available somewhere else? I did a quick search through HGVS and didn't see it, and I wouldn't expect it to be in bioutils, but just making sure

larrybabb replied:
I know. it seems like someone stuck this here to make sure we only worked with "refseq" accessions. I do think that hgvs should retain this mapping structure since it is really an hgvs thing.

larrybabb replied:
this was moved to the dataproxy classes for now. further refactoring will need to happen in another round to allow us to better organize/segregate hgvs code from vrs-python code.

jsstevenson commented on src/ga4gh/vrs/dataproxy.py:
```diff
@@ -88,6 +89,56 @@ def translate_sequence_identifier(self, identifier, namespace=None):
+                f"({start_pos}, {end_pos}) but found {actual_ref}"
+            )
+            _logger.warning(err_msg)
+        return is_valid, err_msg
```
I think the return object could be simplified here. If `is_valid` is always false when `err_msg` is anything other than an empty string, and always true if `err_msg` is an empty string, does it add any additional information? Could we rephrase this to simply check for invalidity, returning `None` if there are no invalidations, and a string description if it is invalid? Alternatively -- maybe raise an exception if invalid and do nothing if valid?

larrybabb replied:
i like your change request here. I am going to make a separate issue for it so as to not open another thread on this PR that is already bigger than I had anticipated.

jsstevenson commented:
It would be nice, albeit out of scope here, if the generic `dataproxy` interface that's redefined here from seqrepo was located in a more central package so that we don't do any damage by making changes in VRS-Python without mirroring them back to SeqRepo.

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -64,39 +56,49 @@ def __init__(
-        sequence = coerce_namespace(sv.ac)
-        return self._get_refget_accession(sequence)
+        prefix_dict = {
+            "refseq:NM_": "n",
```
I think this should be "c"
```suggestion
            "refseq:NM_": "c",
```

larrybabb replied:
on it. I ended up moving some of this code to the seqrepo dataproxy classes.

korikuzma commented:
@larrybabb can you pull in changes from `main` branch and resolve conflicts first?

korikuzma commented on notebooks/variation_normalizer_rest_dp.py:
```diff
@@ -1,5 +1,7 @@
 import requests
 
+""" TODO convert this to a jupyter notebook """
+
```
We can remove the todo and add this to #242 

```suggestion
```

larrybabb replied:
yes

larrybabb commented:
@korikuzma @jsstevenson @ahwagner could you please try to prioritize reviewing this PR? The primary purpose of it was to get support for the `c.` nomenclature working. I also reverted the hgvs branch dependency that we introduced and will localize our version of it until that hgvs branch is officially merged and a new release is created. I started to consolidate all the hgvs-related logic in the hgvs_tools class, but deferred making any larger strategic design changes until we can meet as a team and agree on what and when we need to do there.

korikuzma commented:
@larrybabb I am in the middle of reviewing 😄

korikuzma commented on src/ga4gh/vrs/dataproxy.py:
```diff
@@ -88,6 +119,56 @@ def translate_sequence_identifier(self, identifier, namespace=None):
+        :return: Refget Accession if found
+        """
+
+        if ac == None:
```
https://docs.astral.sh/ruff/rules/none-comparison/

```suggestion
        if ac is None:
```

korikuzma commented on src/ga4gh/vrs/dataproxy.py:
```diff
@@ -126,7 +207,7 @@ def _get_sequence(self, identifier, start=None, end=None):
     def _get_metadata(self, identifier):
         ns, a = coerce_namespace(identifier).split(":", 2)
-        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))
+        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))        
```
```suggestion
        r = list(self.sr.aliases.find_aliases(namespace=ns, alias=a))
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -47,11 +40,12 @@ class Translator:
         re.IGNORECASE
     )
     hgvs_re = re.compile(r"[^:]+:[cgmnpr]\.")
+    hgvs_re = re.compile(r"[^:]+:[cgmnpr]\.")
```
We can remove both since it now lives in `HgvsTools`

```suggestion
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -356,7 +275,8 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
 
         """
-
+        
+        print("gnomad_expr: ", gnomad_expr)
```
```suggestion
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
+        refget_accession = self.data_proxy.derive_refget_accession(sequence)
         if not refget_accession:
             return None
+        
+        print("refget_accession: ", refget_accession)
```
```suggestion
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
+        if kwargs.get("require_validation", True):
+            print("validation check for matching reference sequence bases")
```
```suggestion
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
-        sstate = models.LiteralSequenceExpression(sequence=ins_seq)
-        allele = models.Allele(location=location, state=sstate)
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
```
I have a PR (#375) that simplifies this. I can update it once this PR gets merged.
```suggestion
```

larrybabb replied:
+1

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -366,105 +286,36 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
+                raise ValidationError(err_msg)
+            
+        values = {"refget_accession": refget_accession, "start": start, "end": end, "literal_sequence": ins_seq}
+        print("values: ", values)
```
```suggestion
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -171,32 +115,12 @@ def translate_to(self, vo, fmt):
-
     def _get_hgvs_tools(self):
         """ Only create UTA db connection if needed. There will be one connection per
         translator.
         """
         if self.hgvs_tools is None:
-            uta_conn = hgvs.dataproviders.uta.connect()
-            self.hgvs_tools = HgvsTools(uta_conn)
+            self.hgvs_tools = HgvsTools(self.data_proxy)
         return self.hgvs_tools
```
IMO I think we should use the `lazy_property` decorator 

```python
from ga4gh.vrs.extras.decorators import lazy_property

@lazy_property
def hgvs_tools(self):
    """instantiates and returns an HgvsTools instance"""
    return HgvsTools(self.data_proxy)
```

larrybabb replied:
+1

korikuzma commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,333 @@
+        parser (hgvs.parser.Parser): The HGVS parser object.
+        uta_conn: The UTA (Universal Transcript Archive) connection object.
+        normalizer: The HGVS normalizer object.
+        variantMapper: The HGVS variant mapper object.
```
what about `dp`?

larrybabb replied:
done

korikuzma commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,333 @@
+        self.variantMapper = None
+        self.dp = dp
+
+    def _fetch_uta_conn(self):
+        # just in time instantiation of the UTA connection
+        if self.uta_conn is None:
+            self.uta_conn = hgvs.dataproviders.uta.connect()
+        return self.uta_conn
+    
+    def _fetch_normalizer(self):
+        # just in time instantiation of the normalizer service, with validation set to True
+        if self.normalizer is None:
+            self.normalizer = hgvs.normalizer.Normalizer(self._fetch_uta_conn(), validate=True)
+        return self.normalizer
+    
+    def _fetch_variantMapper(self):
+        # just in time instantiation of the variantMapper service
+        if self.variantMapper is None:
+            self.variantMapper = hgvs.variantmapper.VariantMapper(self._fetch_uta_conn())
+        return self.variantMapper
```
IMO, I think we should just initialize these in the `__init__` method

larrybabb replied:
+1

korikuzma commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,333 @@
         self.parser = hgvs.parser.Parser()
+        self.uta_conn = None
+        self.normalizer = None
+        self.variantMapper = None
```
```suggestion
        self.variant_mapper = None
```

larrybabb replied:
on it

korikuzma commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,333 @@
+    
+
+    def get_edit_type(self, sv: hgvs.sequencevariant.SequenceVariant):
+        if sv == None or sv.posedit == None or sv.posedit.edit == None:
```
```suggestion
        if sv is None or sv.posedit is None or sv.posedit.edit is None:
```

korikuzma commented on src/ga4gh/vrs/utils/hgvs_tools.py:
```diff
@@ -1,26 +1,333 @@
+        if sv.type == "c":
+            sv = self.c_to_n(sv)
+
+        # TODO: shouldn't we be validating the "ref" sequence here? if the kwargs "require_validation" value is set to True
```
This will be added in #151

korikuzma replied:
```suggestion
```

korikuzma reviewed (changes requested):
@larrybabb Great work! Overall, I like this change. I made some suggestions

jsstevenson commented:
@larrybabb as Kori said, I'm out for at least today (which will probably entail some frantic catch-up once I'm back) but for the sake of pragmatism, we can assume my approval once Kori's additional notes above are addressed

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -258,6 +152,28 @@ def __init__(
+                'refget_accession' (str): The accession ID of the reference genome.
+                'start' (int): The start position of the allele.
+                'end' (int): The end position of the allele.
+                literal_sequence' (str): The literal sequence for the allele.
```
```suggestion
                'literal_sequence' (str): The literal sequence for the allele.
```

korikuzma commented on src/ga4gh/vrs/extras/translator.py:
```diff
@@ -376,95 +290,22 @@ def _from_gnomad(self, gnomad_expr, **kwargs):
-        sstate = models.LiteralSequenceExpression(sequence=ins_seq)
-        allele = models.Allele(location=location, state=sstate)
-        allele = self._post_process_imported_allele(allele, **kwargs)
+        # TODO: ask other devs if this should be down on all _from_...  methods?
```
```suggestion
```

korikuzma reviewed (approved):
Two minor things. Good job @larrybabb
